### PR TITLE
Optional Ports for gateway-proxy pods and services

### PR DESCRIPTION
**Pull request by Bslabe123**
# Description

When we create Gloos k8s manifest via helm we expose all ports on gateway-proxy pods/services by default. In certain situations it may be prudent to not allow traffic on certain ports (i.e. HTTPS but no HTTP traffic). There also may be situations in which leaving a port open may lead to unneeded resources being created (as mentioned in the issue, hyperscaler firewall rules). We would like to introduce a toggle/convention to allow anyone building helm charts to disable the exposing of certain ports on gateway-proxy pods/services.

# Context

We have decided to make 0 a reserved value indicating that the port should be disabled, when rendering the helm chart we wrap the port declaration in an if(port != 0)/else statement. We also call this out loudly in the field description since its a less explicit toggle.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by ben-taussig-solo — APPROVED** [review]

File: install/helm/gloo/templates/8-gateway-proxy-service.yaml [hunk 1]
```diff
@@ -34,7 +34,7 @@ spec:
   # port order matters due to this issue: https://github.com/solo-io/gloo/issues/2571
   ports:
 {{- if $spec.service.httpsFirst }}
-  {{- if not $gatewaySettings.disableHttpsGateway }}
+  {{- if and (not $gatewaySettings.disableHttpsGateway) $spec.service.httpsPort }}
```
Comment (sam-heilbron) [inline install/helm/gloo/templates/8-gateway-proxy-service.yaml:37]: Is it the service httpsPort or the podTemplate httpsPort that should be part of the conditional?

File: install/helm/gloo/templates/8-gateway-proxy-service.yaml [hunk 2]
```diff
@@ -139,9 +139,11 @@ spec:
   selector:
     gloo: gateway-proxy
     gateway-proxy-id: {{ $name | kebabcase }}
+{{- if and .Values.global.glooStats.enabled .Values.global.glooStats.serviceMonitorEnabled }}
```
Comment (sam-heilbron) [inline install/helm/gloo/templates/8-gateway-proxy-service.yaml:142]: Is this related to the http(s) port disabling or just an additional change?

**Review by sam-heilbron — CHANGES_REQUESTED** [review]
@jenshu raises a good point about the relationship between `disableHttpGateway` and setting httpPort=0 on gateway-proxy. Why do we need both flags?

File: install/helm/gloo/templates/7-gateway-proxy-deployment.yaml [hunk 3]
```diff
@@ -361,6 +364,7 @@ spec:
           initialDelaySeconds: 3
           periodSeconds: 10
           failureThreshold: 3
+{{- end }}
```
Comment (jenshu) [inline install/helm/gloo/templates/7-gateway-proxy-deployment.yaml:367]: i think this change should be undone

**Comment by Bslabe123** [comment]
See: https://github.com/solo-io/gloo/issues/7377#issuecomment-1387445788

**Pull request closed without merging**